modules/packages/ZMQ.chpl

```diff
@@ -0,0 +1,451 @@
+  private extern const ZMQ_SNDHWM: c_int;
+  private extern const ZMQ_RCVHWM: c_int;
+  private extern const ZMQ_MULTICAST_HOPS: c_int;
+  private extern const ZMQ_RCVTIMEO: c_int;
+  private extern const ZMQ_SNDTIMEO: c_int;
+  private extern const ZMQ_LAST_ENDPOINT: c_int;
+  private extern const ZMQ_ROUTER_MANDATORY: c_int;
+  private extern const ZMQ_TCP_KEEPALIVE: c_int;
+  private extern const ZMQ_TCP_KEEPALIVE_CNT: c_int;
+  private extern const ZMQ_TCP_KEEPALIVE_IDLE: c_int;
+  private extern const ZMQ_TCP_KEEPALIVE_INTVL: c_int;
+  private extern const ZMQ_TCP_ACCEPT_FILTER: c_int;
+  private extern const ZMQ_IMMEDIATE: c_int;
+  private extern const ZMQ_XPUB_VERBOSE: c_int;
+  private extern const ZMQ_ROUTER_RAW: c_int;
+  private extern const ZMQ_IPV6: c_int;
+  private extern const ZMQ_MECHANISM: c_int;
+  private extern const ZMQ_PLAIN_SERVER: c_int;
+  private extern const ZMQ_PLAIN_USERNAME: c_int;
+  private extern const ZMQ_PLAIN_PASSWORD: c_int;
+  private extern const ZMQ_CURVE_SERVER: c_int;
+  private extern const ZMQ_CURVE_PUBLICKEY: c_int;
+  private extern const ZMQ_CURVE_SECRETKEY: c_int;
+  private extern const ZMQ_CURVE_SERVERKEY: c_int;
+  private extern const ZMQ_PROBE_ROUTER: c_int;
+  private extern const ZMQ_REQ_CORRELATE: c_int;
+  private extern const ZMQ_REQ_RELAXED: c_int;
+  private extern const ZMQ_CONFLATE: c_int;
+  private extern const ZMQ_ZAP_DOMAIN: c_int;
+
+  // -- Message Options
+  private extern const ZMQ_MORE: c_int;
+
+  // -- Send/Recv Options
+  private extern const ZMQ_DONTWAIT: c_int;
+  private extern const ZMQ_SNDMORE: c_int;
+
+  // -- Security Options
+  private extern const ZMQ_NULL: c_int;
+  private extern const ZMQ_PLAIN: c_int;
+  private extern const ZMQ_CURVE: c_int;
+
+  /*
+    Query the ZMQ library version.
+
+    :returns: An :type:`(int,int,int)` tuple of the major, minor, and patch
+        version of the ZMQ library.
+   */
+  proc version: (int,int,int) {
+    var major, minor, patch: c_int;
+    zmq_version(c_ptrTo(major), c_ptrTo(minor), c_ptrTo(patch));
+    return (major:int, minor:int, patch:int);
+  }
+
+  /*
+    ZMQ.Context
+
+    The ZMQ Context class maintains a vector of child sockets.  This allows
+    users to delete the Context object without managing the closure and
+    deletion of Sockets created from this Context.
+
+    In ZMQ's C API, context "objects" are thread safe. The lock internal to
+    this Chapel class is only used to serialize accesses to the vector-style
+    array of child sockets.
+  */
+  class Context {
+
+    // Member Elements
+    var ctx: c_void_ptr;
+    var socks: [1..0] Socket;
+    var lock: sync bool;
+
+    // Constructor
+    proc Context() {
+      this.ctx = zmq_ctx_new();
+      if this.ctx == nil {
+        var errmsg = zmq_strerror(errno):string;
+        writef("Error in Context(): %s\n", errmsg);
+      }
+    }
+
+    // Destructor
+    pragma "no doc"
+    proc ~Context() {
+      // Close any open sockets
```

Comment:
I'd add a comment here mentioning that you don't use this.lock because things will fail spectacularly if there is a race between users of a context calling socket and another task destroying it.